src/odemis/gui/comp/overlay/shapes.py

```diff
@@ -92,7 +99,41 @@ def move_to(self, pos: Union[Tuple[float, float], Vec]):
 
     @abstractmethod
     def refresh(self):
-        """Refresh the shape's points VA after copy or move_to operation."""
+        """Refresh the shape's points VA after copy, move_to or restore_state operation."""
+        pass
+
+    def undo(self):
+        """Undo the last action."""
+        if self.undo_stack:
+            state = self.undo_stack.pop()
+            self.redo_stack.append(state)
+            self.restore_state(state)
```

Comment:
Does this work properly? Isn't there some sort of "off by 1" error?
The current state is the "topest" state on the undo_stack, right? Then the first time you call it, nothing will be undone. 
I think you should restore the state to `self.undo_stack[-1]`